### PR TITLE
feat(filters): get filters on /orders & improve filtering + pagination

### PR DESCRIPTION
**Pull request by Mitsichury**

### CODE REVIEW

File: src/converter/index.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { isNumeric } from "../validator/index.js";
+
+export function toStrings(value: string, separator = ","): string[] {
+    return value !== undefined && value !== null && (typeof value == "string") && value.trim() != "" ? value.split(separator) : undefined;
```
Comment (Mitsichury) [inline src/converter/index.ts:4]: return empty

File: src/database/AcebaseClient.ts [hunk 2]
```diff
@@ -22,50 +30,66 @@ export class AceBaseClient implements Database {
...
         return this.db.close()
     }
 
     async addOrder(otcOrder: OtcOrder): Promise<void> {
-        let toAdd = { ...otcOrder, ...otcOrder.order};
+        let toAdd = { ...otcOrder, ...otcOrder.order };
         delete toAdd.order;
         await this.db.ref(ENTRY_REF).push(toAdd);
+        //@ts-ignore
+        this.filters.addSignerToken(otcOrder.order.signerToken, otcOrder.order.signerAmount);
```
Comment (Mitsichury) [inline src/database/AcebaseClient.ts:90]: check every ts ignore

File: src/database/filter/RequestFilter.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import { SortOrder } from './SortOrder';
+
+export class RequestFilter {
+    signerTokens?: string[];
+    senderTokens?: string[];
+    minSignerAmount?: number;
+    maxSignerAmount?: number;
+    minSenderAmount?: number;
+    maxSenderAmount?: number;
+    // @todo: use it
```
Comment (Mitsichury) [inline src/database/filter/RequestFilter.ts:11]: remove it

File: src/database/filter/SortOrder.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export enum SortOrder {
+    ASC ="ASC",
```
Comment (Mitsichury) [inline src/database/filter/SortOrder.ts:2]: format file

File: src/ip_helper.ts [hunk 5]
```diff
@@ -9,9 +9,7 @@ export function getLocalIp() {
       if (net === undefined) {
         continue;
       }
-      // 'IPv4' is in Node <= 17, from 18 it's a number 4 or 6
-      const familyV4Value = typeof net.family === 'string' ? 'IPv4' : 4;
-      if (net.family === familyV4Value && !net.internal) {
+      if (net.family === "IPv4" && !net.internal) {
```
Comment (Mitsichury) [inline src/ip_helper.ts:12]: revert this change

File: src/validator/index.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+export function isDateInRange(date: string, validationDurationInWeek: number) {
+    if (!isNumeric(date)) {
+        return false;
+    }
+
+    let maxDate = new Date();
+    maxDate.setDate(maxDate.getDate() + validationDurationInWeek * 7);
+    console.log(date, maxDate)
```
Comment (Mitsichury) [inline src/validator/index.ts:12]: remove console log

File: src/database/InMemoryDatabase.ts [hunk 7]
```diff
@@ -1,60 +1,113 @@
...
+      return Promise.resolve(new OrderResponse(result, 1));
+    }
+    return Promise.resolve(new OrderResponse(null, 0));
   }
 
-  async getOrders(): Promise<Record<string, OtcOrder>> {
-    return Promise.resolve(this.database);
+  async getOrders(): Promise<OrderResponse> {
+    const size = Object.keys(this.database).length;
+    console.log(size);
```
Comment (Mitsichury) [inline src/database/InMemoryDatabase.ts:102]: remove log

File: src/database/filter/RequestFilter.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    page?: number;
+    sortField?: SortField;
+    sortOrder?: SortOrder;
+    maxAddedDate?: number;
+}
+
+
```
Comment (Mitsichury) [inline src/database/filter/RequestFilter.ts:18]: clean file

File: benchmarking/bench.js [hunk 9]
```diff
@@ -1,5 +1,7 @@
 import axios from "axios";
 
+const tokens = ["BTC", "ETH", "DAI", "AST"];
```
Comment (dmosites) [inline benchmarking/bench.js:3]: Tokens will always be addresses so suggest we use those for testing.

File: src/mapper/mapAnyToOrder.ts [hunk 10]
```diff
@@ -1,14 +1,19 @@
...
+import { DbOrder } from 'model/DbOrder.js';
+import { toNumber } from '../converter/index.js';
+
+export function mapAnyToDbOrder(data: any): DbOrder {
+    if (!data || typeof data != 'object') {
+        return undefined;
+    }
 
-export function mapAnyToOrder(data: any): Order {
     return {
```
Comment (dmosites) [inline src/mapper/mapAnyToOrder.ts:9]: All values should be strings because JavaScript has problems handling big numbers. Either the database should parse into a number internally or BigNumber should be used here.

**Review by piersss — CHANGES_REQUESTED** [review]

File: src/database/filter/__tests__/Filters.test.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+describe("Filters", () => {
+
+    describe("sender Token", () => {
+        test("add new senderToken", () => {
+            const filters = new Filters();
+    
+            filters.addSenderToken("eth", 10);
+    
+            expect(filters.senderToken).toEqual({ "eth": { min: 10, max: 10 } });
+        });
```
Comment (piersss) [inline src/database/filter/__tests__/Filters.test.ts:11]:
```suggestion
describe("Filters", () => {
    let filters: Filters;

    beforeEach(() => {
        filters = new Filters();
    });

    describe("sender Token", () => {
        test("add new senderToken", () => {
            filters.addSenderToken("eth", 10);
    
            expect(filters.senderToken).toEqual({ "eth": { min: 10, max: 10 } });
        });
```

Comment (piersss) [reply src/database/filter/__tests__/Filters.test.ts:11]: This way you can remove all the `const filters = new Filters();` in your tests.

File: src/database/filter/Filters.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+    addSignerToken(token: string, amount: number): void {
+        this.editTokenMap(this.signerToken, token, amount);
+    }
+
+    addSenderToken(token: string, amount: number): void {
+        this.editTokenMap(this.senderToken, token, amount);
+    }
+
+    private editTokenMap(map: Record<string, AmountLimitFilter>, token: string, amount: number): void {
```
Comment (piersss) [inline src/database/filter/Filters.ts:20]: Add a check here if the amount is a positive number. Also token should not be case sensitive, so add a `toLowerCase()` here.

**Review by piersss — CHANGES_REQUESTED** [review]

File: src/database/InMemoryDatabase.ts [hunk 13]
```diff
@@ -1,60 +1,102 @@
...
+            return (b.order.signerAmount) - (a.order.signerAmount)
+          }
+        }
+        if (requestFilter.sortField == SortField.SENDER_AMOUNT) {
+          if (requestFilter.sortOrder == SortOrder.ASC) {
+            return (a.order.senderAmount) - (b.order.senderAmount)
+          } else {
+            return (b.order.senderAmount) - (a.order.senderAmount)
+          }
+        }
```
Comment (piersss) [inline src/database/InMemoryDatabase.ts:50]:
```suggestion
        if (requestFilter.sortOrder == SortOrder.ASC) {
          return (a.order.senderAmount) - (b.order.senderAmount)
        }

        return (b.order.senderAmount) - (a.order.senderAmount);
```

**Review by piersss — CHANGES_REQUESTED** [review]

File: src/database/filter/SortField.ts [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+export function toSortField(key: string) : SortField {
+    if (key === null || key === undefined || !(typeof key === 'string')) {
+        return undefined;
+      }
+      const upperCaseValue = key.toUpperCase();
+      const match = Object.keys(SortField).filter(s => s === upperCaseValue);
+      return match.length === 1 ? match[0] as SortField : undefined;
+}
```
Comment (piersss) [inline src/database/filter/SortField.ts:13]:
```suggestion
export function toSortField(key: string) : SortField | undefined {
    if (typeof key !== 'string') {
        return undefined;
    }

    if (key.toUpperCase() === SortField.SIGNER_AMOUNT) {
        return SortField.SIGNER_AMOUNT;
    }

    if (key.toUpperCase() === SortField.SENDER_AMOUNT) {
        return SortField.SENDER_AMOUNT;
    }

    return undefined;
}

```

File: src/database/filter/SortOrder.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+export function toSortOrder(key: string): SortOrder {
+    if (key === null || key === undefined || !(typeof key === 'string')) {
+        return undefined;
+    }
+    const upperCaseValue = key.toUpperCase();
+    const match = Object.keys(SortOrder).filter(s => s === upperCaseValue);
+    return match.length === 1 ? match[0] as SortOrder : undefined;
+}
```
Comment (piersss) [inline src/database/filter/SortOrder.ts:13]:
```suggestion
export function toSortOrder(key: string): SortOrder | undefined {
    if (typeof key !== 'string') {
        return undefined;
    }

    if (key.toUpperCase() === SortOrder.ASC) {
        return SortOrder.ASC;
    }

    if (key.toUpperCase() === SortOrder.DESC) {
        return SortOrder.DESC;
    }
    
    return undefined;
}
```

**Review by piersss — CHANGES_REQUESTED** [review]
Suggested some small changes. :)

I was wondering why you're not using typescript in strict mode? It's very useful!

File: src/database/filter/__tests__/Filters.test.ts (hunk 11 above)
Comment (Mitsichury) [reply src/database/filter/__tests__/Filters.test.ts]: Can do, but it breaks the pattern Arrange/Act/Assert

File: src/model/OtcOrder.ts [hunk 16]
```diff
@@ -3,9 +3,9 @@ import { Order } from '@airswap/typescript';
 export class OtcOrder {
```
[the remark this answers is not in the document]
Comment (Mitsichury) [reply src/model/OtcOrder.ts]: Fixed

File: src/model/OtcOrder.ts [hunk 17]
```diff
@@ -3,9 +3,9 @@ import { Order } from '@airswap/typescript';
...
     id: string;
```
[the remark this answers is not in the document]
Comment (Mitsichury) [reply src/model/OtcOrder.ts]: Changed, I use it as an id for the database, could cost a lot to compute it multiple times

File: benchmarking/bench.js (hunk 9 above)
Comment (Mitsichury) [reply benchmarking/bench.js]: Changed

File: src/mapper/mapAnyToOrder.ts (hunk 10 above)
Comment (Mitsichury) [reply src/mapper/mapAnyToOrder.ts:9]: Fixed !
All strings for ui/smart contract, I added an approximative field in order to sort and limit

File: src/database/filter/__tests__/Filters.test.ts (hunk 11 above)
Comment (Mitsichury) [reply src/database/filter/__tests__/Filters.test.ts]: done

File: src/database/InMemoryDatabase.ts (hunk 13 above)
Comment (Mitsichury) [reply src/database/InMemoryDatabase.ts]: done

File: src/model/OrderResponse.ts [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+    orders: Record<string, OtcOrder>;
+    totalPages: number;
+    filters: Filters;
+
+    constructor(orders: Record<string, OtcOrder>, totalPages: number, filters?: Filters) {
+        this.orders = orders;
+        this.totalPages = totalPages;
+        this.filters = filters;
+    }
+}
```
[the remark this answers is not in the document]
Comment (Mitsichury) [reply src/model/OrderResponse.ts]: done !

File: src/database/filter/Filters.ts (hunk 12 above)
Comment (Mitsichury) [reply src/database/filter/Filters.ts:20]: Done !

File: src/database/filter/SortOrder.ts (hunk 15 above)
Comment (Mitsichury) [reply src/database/filter/SortOrder.ts]: changed

File: src/database/filter/SortField.ts (hunk 14 above)
Comment (Mitsichury) [reply src/database/filter/SortField.ts]: changed

File: src/database/InMemoryDatabase.ts [hunk 19]
```diff
@@ -1,60 +1,102 @@
...
+      });
+    const totalResultsCount = totalResults.length;
+    totalResults.slice((requestFilter.page - 1) * elementPerPage, requestFilter.page * elementPerPage)
+      .forEach((OtcOrder) => {
+        const orderId = OtcOrder['id'];
+        orders[`${orderId}`] = OtcOrder;
+      });
```
[the remark this answers is not in the document]
Comment (Mitsichury) [reply src/database/InMemoryDatabase.ts]: much better !

File: src/model/OrderResponse.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (piersss) [reply src/model/OrderResponse.ts]: Nice. but please keep `totalPages`. That's not included in Pagination.

**Comment by piersss** [comment]
Resolved changes, just one small thing about the pagination.

File: src/model/OrderResponse.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+    orders: Record<string, IndexedOrder>;
+    pagination: Pagination;
+    filters: Filters | undefined;
+    entriesForQuery: number;
+
+    constructor(orders: Record<string, IndexedOrder>, pagination: Pagination, entriesForQuery: number, filters?: Filters) {
+        this.orders = orders;
+        this.pagination = pagination;
+        this.filters = filters;
+        this.entriesForQuery = entriesForQuery;
```
Comment (piersss) [inline src/model/OrderResponse.ts:15]: `entriesForQuery` means total orders?

**Review by piersss — APPROVED** [review]

**Comment by gpxl-dev** [comment]
Is there a benefit to having an explicit `filters` param rather than implying its value based on whether any of the actual filter params are passed?
